### PR TITLE
[Bugfix][Habana_main] fix guided_decode HPU failing issue

### PR DESCRIPTION
**Pull request by xuechendi**
FILL IN THE PR DESCRIPTION HERE

FIX #https://github.com/HabanaAI/vllm-fork/issues/198

After this change, we can see tool_calls can be returned successfully

``` bash
Compiling FSM index for all state transitions: 100%|████████████████████████████████████████████████████████████████████████| 55/55 [00:01<00:00, 32.86it/s]INFO 09-04 02:15:34 metrics.py:406] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 0 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.0%, CPU KV cache usage: 0.0%.
INFO 09-04 02:15:34 logger.py:36] Received request chat-0fd03b03ae05473488d9bce566401d91: prompt: "<|im_start|>user\nWhat's the weather like in Boston today?<|im_end|>\n<|im_start|>assistant\n", params: SamplingParams(n=1, best_of=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=0.7, top_p=1.0, top_k=-1, min_p=0.0, seed=None, use_beam_search=False, length_penalty=1.0, early_stopping=False, stop=[], stop_token_ids=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=1000, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None), prompt_token_ids: [27, 91, 318, 5011, 91, 29, 882, 198, 3923, 596, 279, 9282, 1093, 304, 10406, 3432, 76514, 91, 318, 6345, 91, 397, 27, 91, 318, 5011, 91, 29, 78191, 198], lora_request: None, prompt_adapter_request: None.
INFO 09-04 02:15:34 async_llm_engine.py:173] Added request chat-0fd03b03ae05473488d9bce566401d91.
INFO 09-04 02:15:36 async_llm_engine.py:140] Finished request chat-0fd03b03ae05473488d9bce566401d91.
INFO:     127.0.0.1:40452 - "POST /v1/chat/completions HTTP/1.1" 200 OK
Message: ChatCompletionMessage(content='', refusal=None, role='assistant', function_call=None, tool_calls=[ChatCompletionMessageToolCall(id='chatcmpl-tool-af3eac9372144f959ed0df7e16cf5da4', function=Function(arguments='{ "location": "Boston, MA", "unit": "fahrenheit" }', name='get_current_weather'), type='function')])
```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by xuechendi** [comment]
@kzawora-intel , please help to review this PR. This is a bug fix for issue #198

**Comment by xuechendi** [comment]
test script as below

``` python 
# VLLM_SKIP_WARMUP=true PT_HPU_ENABLE_LAZY_COLLECTIVES=true VLLM_RAY_DISABLE_LOG_TO_DRIVER=1

import asyncio, os, sys, subprocess, time
from contextlib import contextmanager
import openai
from openai._types import NOT_GIVEN
import requests
from typing import Any, Callable, Dict, List, Optional
from vllm.entrypoints.openai.cli_args import make_arg_parser
from vllm.utils import FlexibleArgumentParser, get_open_port
from pathlib import Path

VLLM_PATH = os.path.join(Path(__file__).parent.parent, "vllm")
sys.path.append(VLLM_PATH)
"""Path to root of the vLLM repository."""

@contextmanager
def _nvml():
    yield
        
class RemoteOpenAIServer:
    DUMMY_API_KEY = "token-abc123"  # vLLM's OpenAI server does not need API key

    def __init__(self,
                 model: str,
                 cli_args: List[str],
                 *,
                 env_dict: Optional[Dict[str, str]] = None,
                 auto_port: bool = True,
                 max_wait_seconds: Optional[float] = None) -> None:
        if auto_port:
            if "-p" in cli_args or "--port" in cli_args:
                raise ValueError("You have manually specified the port"
                                 "when `auto_port=True`.")

            cli_args = cli_args + ["--port", str(get_open_port())]

        parser = FlexibleArgumentParser(
            description="vLLM's remote OpenAI server.")
        parser = make_arg_parser(parser)
        args = parser.parse_args(cli_args)
        #print(args)
        self.host = str(args.host or 'localhost')
        self.port = int(args.port)

        env = os.environ.copy()
        # the current process might initialize cuda,
        # to be safe, we should use spawn method
        env['VLLM_WORKER_MULTIPROC_METHOD'] = 'spawn'
        if env_dict is not None:
            env.update(env_dict)
        self.proc = subprocess.Popen(["vllm", "serve"] + [model] + cli_args,
                                     env=env,
                                     stdout=sys.stdout,
                                     stderr=sys.stderr)
        max_wait_seconds = max_wait_seconds or 240
        self._wait_for_server(url=self.url_for("health"),
                              timeout=max_wait_seconds)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.proc.terminate()
        try:
            self.proc.wait(3)
        except subprocess.TimeoutExpired:
            # force kill if needed
            self.proc.kill()

    def _wait_for_server(self, *, url: str, timeout: float):
        # run health check
        start = time.time()
        while True:
            try:
                if requests.get(url).status_code == 200:
                    break
            except Exception as err:
                result = self.proc.poll()
                if result is not None and result != 0:
                    raise RuntimeError("Server exited unexpectedly.") from err

                time.sleep(0.5)
                if time.time() - start > timeout:
                    raise RuntimeError(
                        "Server failed to start in time.") from err

    @property
    def url_root(self) -> str:
        return f"http://{self.host}:{self.port}"

    def url_for(self, *parts: str) -> str:
        return self.url_root + "/" + "/".join(parts)

    def get_client(self):
        return openai.OpenAI(
            base_url=self.url_for("v1"),
            api_key=self.DUMMY_API_KEY,
        )

    def get_async_client(self):
        return openai.AsyncOpenAI(
            base_url=self.url_for("v1"),
            api_key=self.DUMMY_API_KEY,
        )

async def test_named_tool_use(client: openai.AsyncOpenAI, model_name, input, tools, tool_choice):
    # non-streaming
    chat_completion = await client.chat.completions.create(
        model=model_name,
        messages=input,
        max_tokens=1000,
        tools=tools,
        tool_choice=tool_choice)
    message = chat_completion.choices[0].message
    print(f"Message: {message}")

def test_multiple_sampling_params():
    
    model_name = "meta-llama/Meta-Llama-3.1-8B"
    args = [
        "--max-model-len", "8192",
        #"--enforce-eager",
    ]
    with RemoteOpenAIServer(model_name, args) as remote_server:
        
        client_inst = remote_server.get_async_client()
    
        input = [{"role": "user", "content": "What's the weather like in Boston today?"}]
        tools = [{
            "type": "function",
            "function": {
                "name": "get_current_weather",
                "description": "Get the current weather in a given location",
                "parameters": {
                    "type": "object",
                    "properties": {
                    "location": {
                        "type": "string",
                        "description": "The city and state, e.g. San Francisco, CA",
                    },
                    "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
                    },
                    "required": ["location"],
                },
            }
        }]
        tool_choice = {'function': {'name': 'get_current_weather'}}
        #tool_choice = NOT_GIVEN
        
        asyncio.run(test_named_tool_use(client_inst, model_name, input, tools, tool_choice=tool_choice))
        
if __name__ == "__main__":
    print("VLLM_PATH is ", VLLM_PATH)
    test_multiple_sampling_params()
```

**Comment by tae-su-kim** [comment]
Duplicate of PR #153 and PR #226 - please review those PRs as well.

**Comment by xuechendi** [comment]
@kzawora-intel , I noticed that original in-place tensor.add_ is not well supported in HPU. One easy fix is to change it to 'add'.

**Comment by xuechendi** [comment]
> Duplicate of PR #153 and PR #226 - please review those PRs as well.

Thanks, @tae-su-kim , yeah, either 226 or this PR works for me, I want to make vllm-fork to support Agent tool_calls ASAP so it can be utilized in OPEA.

**Comment by tae-su-kim** [comment]
@xuechendi Great! We recently observed unexpected throughput degradation with guided_decode and submitted fix for it (commit 6d57c18 and https://github.com/HabanaAI/vllm-fork/pull/226#issuecomment-2337291551). If you are interested, please check it out. It would be really helpful if you could cross-check latency improvement with the test cases for tool_call and llama-3.1-8B.

**Review by michalkuligowski — COMMENTED** [review]
LGTM

**Review by michalkuligowski — APPROVED** [review]

**Comment by tae-su-kim** [comment]
@michalkuligowski This feature seriously needs commit [6d57c18](https://github.com/HabanaAI/vllm-fork/pull/226/commits/6d57c1882767c730fb40208a7db4551f3086c873). Without it, e2e throughput degrades a lot. Please review #226 also.

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def __call__(self, input_ids: List[int],
                           -math.inf,
                           device=scores.device)
         mask[allowed_tokens] = 0
-        scores.add_(mask)
+        scores = scores.add(mask)
```
Comment (jikunshang) [inline vllm/model_executor/guided_decoding/outlines_logits_processors.py:64]: just curious, what's the difference here?

Comment (xuechendi) [reply vllm/model_executor/guided_decoding/outlines_logits_processors.py:64]: should be no difference in result, however, 'add_' is executing in_place, while 'add' will return a tensor.
From my test, using 'add_' leads to "RuntimeError: synNodeCreateWithId failed for node: strided_insert with synStatus 1 [Invalid argument]. ", replacing with 'add' fixed above issue

**Pull request merged**